### DISCUSSION TITLE
Atomic migrations?

### DISCUSSION THREAD

**Discussion by gd87429** [SeaQL/sea-orm#1059] [General]
If you run `Migrator::up()` in parallel (like when you spin up multiple servers), is there a chance the same migration might run more than once?

If yes -- are there plans to build-in support for atomic migrations, or should the user make sure migration cannot be applied twice (like checking if data exists before seeding the table)?

**Comment by billy1624** [comment]
Hey @gd87429, hi again! I think the atomic migration is database specific? For example, in MySQL, the DDL statement such as `CREATE TABLE` is not transactional that means it can't be rollback.

I'd love to support atomic migrations but seems it's not a easy task. As we need to come up with a plan for each of our supported database, unless I miss something.

What's your suggestions on that?

**Reply by gd87429** [reply]
Hey @billy1624, yes that's right - it seems it's not an easy task.

One way of going about it could be to have a single place (like a table with one row, or a global variable) that could act as a mutex. Before a migration starts, it could update the lock to **on**, and on finishing all migrations (or until it encounters the first error) set the lock to **off**. But then you'd introduce a new set of problems - like when another program is about to run migrations and encounters a lock, does it wait for it to become unlocked or does it simply skip running its own migrations?

Doesn't seem like the cleanest solution, so I'd probably lean towards holding off until more people weigh in.

Thank you for taking the time tho.

---

**Edit:** For anyone else stumbling on the same issue, rn I'm going around this using `if_not_exists` (table or index creation) and `on_conflict / do_nothing` (seeding data). Doesn't help with every use case tho (like when there are no indexes on table).

**Comment by tyt2y3** [comment]
The general recommendation is that you call https://docs.rs/sea-orm-migration/latest/sea_orm_migration/manager/struct.SchemaManager.html#method.get_connection and then open a transaction on it

**Reply by Mastermindaxe** [reply]
Could you provide an example on how to do this with `Migrator::up`? I can't really wrap my head around it...

**Comment by tyt2y3** [comment]
What is doable but require more effort is for `SchemaManager` to become a enum of `DatabaseConnection` and `DatabaseTransaction`,
and allowing one to open a transaction if on Postgres:

I imagine:

```rust
pub enum SchemaManager<'c> {
    DatabaseConnection(&'c DatabaseConnection),
    DatabaseTransaction(DatabaseTransaction),
}

// and thus

impl SchemaManager {
    fn get_transaction(&self) -> Self {
        if postgres {
        match self {
            DatabaseConnection(conn) => Self::DatabaseTransaction(conn.begin()),
            DatabaseTransaction(tran) => Self::DatabaseTransaction(tran.begin()), // this will open a nested transaction
        }
        } else {
            self.clone()
        }
    }
}
```

Just an imagination, I have not tested whether it will pass all the lifetime requirements.

**Comment by billy1624** [comment]
Hey @gd87429, tracking issue https://github.com/SeaQL/sea-orm/issues/1378

**Comment by RenovZ** [comment]
I think maybe we can try another way to just make sure the migration will be executed successfully. Like dryrun the migraiton, it will not update the database actually just check if the migration itself is ok.
Maybe this will need a PR for mysql, postgres, or sqlite (I'm not sure. LOL).

**Comment by hmacias-avaya** [comment]
I'm using yugabyte which should be close to postgres and the migrations don't seem to be running inside a transaction.
I have several "create index" in a migration and my application was restarted in the middle of a migration.
I now have some indices created but not all of them, and the migration does not run again so they are never created.

Is this expected using the latest seaorm?

also, looking at https://www.sea-ql.org/SeaORM/docs/migration/writing-migration/#atomic-migration it states _You **can** start a transaction inside each migration..._

I know I **can** start a transaction, but do I have to (if I'm using postgres), if the previous paragraph said migrations run inside a transaction anyway?

**Reply by Huliiiiii** [reply]
We wraps migrations in a transaction on Postgres by default. 
This behavior can be manually overridden. See: [https://github.com/SeaQL/sea-orm/pull/2980](https://github.com/SeaQL/sea-orm/pull/2980)

**Reply by hmacias-avaya** [reply]
thanks!

In case anybody else is using yugabyte, I was also affected by the DDLs not being transaction at yugabyte by default (see https://docs.yugabyte.com/stable/explore/transactions/transactional-ddl/)
